libvast/src/format/csv.cpp

```diff
@@ -129,4 +140,271 @@ const char* writer::name() const {
+      return record_type{{t.name(), t}}.name(t.name());
+    } // else skip
+  }
+  return caf::none;
+}
+
+namespace {
+
+template <class Iterator, class Attribute>
+struct container_parser_builder {
+  using result_type = rule<Iterator, Attribute>;
+
+  explicit container_parser_builder(const std::string& set_separator)
+    : set_separator_{set_separator} {
+    // nop
+  }
+
+  // TODO: support enumeration_type inside of containers too.
+  template <class T>
+  result_type operator()(const T& t) const {
+    if constexpr (std::is_same_v<T, string_type>) {
+      return +(parsers::any - set_separator_ - '=')->*[](std::string&& x) {
+        return data{std::move(x)};
+      };
+    } else if constexpr (std::is_same_v<T, pattern_type>) {
+      return +(parsers::any - set_separator_ - '=')->*[](std::string&& x) {
+        return data{pattern{std::move(x)}};
+      };
+    } else if constexpr (std::is_same_v<T, set_type>) {
+      auto set_insert = [](std::vector<Attribute>&& xs) {
+        return set(std::make_move_iterator(xs.begin()),
+                   std::make_move_iterator(xs.end()));
+      };
+      return ('{'_p >> (caf::visit(*this, t.value_type) % set_separator_)
+              >> '}'_p)
+               ->*set_insert;
+    } else if constexpr (std::is_same_v<T, vector_type>) {
+      auto vector_insert = [](std::vector<Attribute>&& xs) { return xs; };
+      return ('[' >> (caf::visit(*this, t.value_type) % set_separator_) >> ']')
+               ->*vector_insert;
+    } else if constexpr (std::is_same_v<T, map_type>) {
+      auto map_insert = [](std::vector<std::tuple<Attribute, Attribute>>&& xs) {
+        auto to_pair = [](auto&& tuple) {
+          return std::make_pair(std::get<0>(tuple), std::get<1>(tuple));
+        };
+        map m;
+        for (auto& x : xs)
+          m.insert(to_pair(std::move(x)));
+        return m;
+      };
+      auto kvp = caf::visit(*this, t.key_type) >> '='
+                 >> caf::visit(*this, t.value_type);
+      return (('{'_p >> (kvp % set_separator_) >> '}')->*map_insert);
+    } else if constexpr (has_parser_v<type_to_data<T>>) {
+      using value_type = type_to_data<T>;
+      auto ws = ignore(*parsers::space);
+      return (ws >> make_parser<value_type>{} >> ws)->*[](value_type&& x) {
+        return x;
+      };
+    } else {
+      VAST_ERROR_ANON("csv parser builder faild to fetch a parser for type",
+                      caf::detail::pretty_type_name(typeid(T)));
+      return {};
+    }
+  }
+
+  std::string set_separator_;
+};
+
+template <class Iterator>
+struct csv_parser_factory {
+  using result_type = type_erased_parser<Iterator>;
+
+  csv_parser_factory(const std::string& set_separator,
+                     table_slice_builder_ptr bptr)
+    : set_separator_{set_separator}, bptr_{std::move(bptr)} {
+    // nop
+  }
+
+  struct identity {
+    template <class T>
+    constexpr T operator()(T&& t) const noexcept {
+      return std::forward<T>(t);
+    }
+  };
```

Comment:
Per our discussion, I think we can get rid of `identity` if we push the functionality in action parsers.